File: javascript/large_function.js
/**
 * JavaScript test file with a large function.
 * 
 * This file contains a large function that will later be broken down
 * into smaller functions to test DiffScope's ability to detect
 * function extraction and composition.
 */

const fs = require('fs');
const path = require('path');

/**
 * Process data from input file, analyze it, and write results to output file.
 * 
 * This function handles the entire data processing pipeline in one large function.
 * It reads data from various file formats, performs analysis, generates statistics,
 * handles errors, and writes the results to a specified output file.
 * 
 * @param {string} inputPath - Path to input file (CSV, JSON, or TXT)
 * @param {string} outputPath - Path to output file (JSON)
 * @param {Object} config - Configuration dictionary with processing options
 * @returns {Object} Object with processing results and metadata
 */
function processAndAnalyzeData(inputPath, outputPath, config) {
  const results = {
    status: "success",
    inputFile: inputPath,
    outputFile: outputPath,
    recordsProcessed: 0,
    errors: [],
    warnings: [],
    statistics: {}
  };
  
  // Validate inputs
  if (!fs.existsSync(inputPath)) {
    results.status = "error";
    results.errors.push(`Input file does not exist: ${inputPath}`);
    return results;
  }
  
  if (typeof config !== 'object' || config === null) {
    results.status = "error";
    results.errors.push("Configuration must be an object");
    return results;
  }
  
  // Extract configuration options
  const columnMapping = config.columnMapping || {};
  const filterCriteria = config.filterCriteria || {};
  const aggregationFields = config.aggregationFields || [];
  const includeStats = config.includeStatistics !== undefined ? config.includeStatistics : true;
  const outputFormat = config.outputFormat || "json";
  
  // Determine file type from extension
  const extension = path.extname(inputPath).toLowerCase();
  
  // Read and parse the input file
  let data = [];
  try {
    if (extension === '.csv') {
      const fileContent = fs.readFileSync(inputPath, 'utf8');
      const lines = fileContent.split('\n').filter(line => line.trim());
      const headers = lines[0].split(',').map(header => header.trim());
      
      for (let i = 1; i < lines.length; i++) {
        const values = lines[i].split(',').map(value => value.trim());
        if (values.length === headers.length) {
          const row = {};
          for (let j = 0; j < headers.length; j++) {
            row[headers[j]] = values[j];
          }
          data.push(row);
        } else {
          results.warnings.push(`Skipped malformed line: ${lines[i]}`);
        }
      }
      results.fileType = "csv";
    } else if (extension === '.json') {
      const fileContent = fs.readFileSync(inputPath, 'utf8');
      data = JSON.parse(fileContent);
      results.fileType = "json";
    } else if (extension === '.txt') {
      const fileContent = fs.readFileSync(inputPath, 'utf8');
      const lines = fileContent.split('\n').filter(line => line.trim());
      const headers = lines[0].split(',').map(header => header.trim());
      
      for (let i = 1; i < lines.length; i++) {
        const values = lines[i].split(',').map(value => value.trim());
        if (values.length === headers.length) {
          const row = {};
          for (let j = 0; j < headers.length; j++) {
            row[headers[j]] = values[j];
          }
          data.push(row);
        } else {
          results.warnings.push(`Skipped malformed line: ${lines[i]}`);
        }
      }
      results.fileType = "txt";
    } else {
      results.status = "error";
      results.errors.push(`Unsupported file extension: ${extension}`);
      return results;
    }
  } catch (error) {
    results.status = "error";
    results.errors.push(`Error reading input file: ${error.message}`);
    return results;
  }
  
  // Ensure data is an array
  if (!Array.isArray(data)) {
    if (typeof data === 'object' && data !== null) {
      data = [data];
    } else {
      results.status = "error";
      results.errors.push("Data must be an array or object");
      return results;
    }
  }
  
  // Apply column mapping if provided
  if (Object.keys(columnMapping).length > 0) {
    const mappedData = [];
    for (const item of data) {
      const mappedItem = {};
      
      // Apply the mapping
      for (const [oldKey, newKey] of Object.entries(columnMapping)) {
        if (Object.prototype.hasOwnProperty.call(item, oldKey)) {
          mappedItem[newKey] = item[oldKey];
        }
      }
      
      // Copy unmapped fields
      for (const [key, value] of Object.entries(item)) {
        if (!Object.prototype.hasOwnProperty.call(columnMapping, key)) {
          mappedItem[key] = value;
        }
      }
      
      mappedData.push(mappedItem);
    }
    data = mappedData;
  }
  
  // Apply filtering if criteria provided
  if (Object.keys(filterCriteria).length > 0) {
    const filteredData = [];
    
    for (const item of data) {
      let include = true;
      
      for (const [field, criteria] of Object.entries(filterCriteria)) {
        if (Object.prototype.hasOwnProperty.call(item, field)) {
          let value = item[field];
          
          // Try to convert numeric strings to numbers for comparison
          if (typeof value === 'string') {
            if (!isNaN(value) && value.trim() !== '') {
              value = value.includes('.') ? parseFloat(value) : parseInt(value, 10);
            }
          }
          
          // Apply different types of criteria
          if (typeof criteria === 'object' && criteria !== null) {
            if (Object.prototype.hasOwnProperty.call(criteria, "eq") && value !== criteria.eq) {
              include = false;
              break;
            }
            if (Object.prototype.hasOwnProperty.call(criteria, "ne") && value === criteria.ne) {
              include = false;
              break;
            }
            if (Object.prototype.hasOwnProperty.call(criteria, "gt") && 
                !(typeof value === 'number' && value > criteria.gt)) {
              include = false;
              break;
            }
            if (Object.prototype.hasOwnProperty.call(criteria, "lt") && 
                !(typeof value === 'number' && value < criteria.lt)) {
              include = false;
              break;
            }
            if (Object.prototype.hasOwnProperty.call(criteria, "in") && 
                !Array.isArray(criteria.in) && !criteria.in.includes(value)) {
              include = false;
              break;
            }
            if (Object.prototype.hasOwnProperty.call(criteria, "contains") && 
                !(typeof value === 'string' && value.includes(criteria.contains))) {
              include = false;
              break;
            }
          } else {
            // Simple equality check
            if (value !== criteria) {
              include = false;
              break;
            }
          }
        } else {
          // Field not found
          include = false;
          break;
        }
      }
      
      if (include) {
        filteredData.push(item);
      }
    }
    
    results.filteredCount = filteredData.length;
    results.originalCount = data.length;
    data = filteredData;
  }
  
  // Process numeric fields and calculate aggregations
  if (data.length > 0 && aggregationFields.length > 0) {
    const aggregatedData = {};
    
    for (const field of aggregationFields) {
      const numericValues = [];
      
      for (const item of data) {
        if (Object.prototype.hasOwnProperty.call(item, field)) {
          let value = item[field];
          
          // Try to convert string to number
          if (typeof value === 'string') {
            if (!isNaN(value) && value.trim() !== '') {
              value = value.includes('.') ? parseFloat(value) : parseInt(value, 10);
            } else {
              continue;
            }
          }
          
          if (typeof value === 'number' && !isNaN(value)) {
            numericValues.push(value);
          }
        }
      }
      
      if (numericValues.length > 0) {
        const sum = numericValues.reduce((acc, val) => acc + val, 0);
        const avg = sum / numericValues.length;
        
        aggregatedData[field] = {
          count: numericValues.length,
          sum: sum,
          average: avg,
          min: Math.min(...numericValues),
          max: Math.max(...numericValues)
        };
        
        // Add more statistics if more than one value
        if (numericValues.length > 1) {
          // Sort for median calculation
          numericValues.sort((a, b) => a - b);
          const midIndex = Math.floor(numericValues.length / 2);
          const median = numericValues.length % 2 === 0
            ? (numericValues[midIndex - 1] + numericValues[midIndex]) / 2
            : numericValues[midIndex];
          
          // Calculate variance and standard deviation
          const variance = numericValues.reduce((acc, val) => acc + Math.pow(val - avg, 2), 0) / numericValues.length;
          const stddev = Math.sqrt(variance);
          
          aggregatedData[field].median = median;
          aggregatedData[field].stddev = stddev;
          aggregatedData[field].variance = variance;
        }
      }
    }
    
    results.aggregations = aggregatedData;
  }
  
  // Compute overall statistics
  if (includeStats && data.length > 0) {
    const fieldTypes = {};
    for (const item of data) {
      for (const [key, value] of Object.entries(item)) {
        if (!Object.prototype.hasOwnProperty.call(fieldTypes, key)) {
          fieldTypes[key] = { type: typeof value, count: 1 };
        } else {
          fieldTypes[key].count += 1;
        }
      }
    }
    
    results.statistics.fieldTypes = fieldTypes;
    results.statistics.recordCount = data.length;
    results.statistics.fieldCount = Object.keys(fieldTypes).length;
  }
  
  // Prepare output
  const outputData = {
    metadata: results,
    data: data
  };
  
  // Write to output file
  try {
    if (outputFormat.toLowerCase() === "json") {
      fs.writeFileSync(outputPath, JSON.stringify(outputData, null, 2));
    } else {
      results.status = "error";
      results.errors.push(`Unsupported output format: ${outputFormat}`);
      return results;
    }
  } catch (error) {
    results.status = "error";
    results.errors.push(`Error writing output file: ${error.message}`);
    return results;
  }
  
  results.recordsProcessed = data.length;
  return results;
}

// Export the function
module.exports = {
  processAndAnalyzeData
};

// If running as a script
if (require.main === module) {
  // Test with sample data
  const config = {
    columnMapping: { name: "fullName", age: "years" },
    filterCriteria: { years: { gt: 20 } },
    aggregationFields: ["years", "score"],
    includeStatistics: true,
    outputFormat: "json"
  };
  
  // Create a sample CSV file for testing
  const sampleCsv = "tests/sample_data.csv";
  const csvContent = "name,age,score,active\nAlice,25,95,true\nBob,30,85,true\nCharlie,20,75,false";
  
  try {
    // Ensure directory exists
    const dir = path.dirname(sampleCsv);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    
    fs.writeFileSync(sampleCsv, csvContent);
    
    const outputJson = "tests/output_data.json";
    const results = processAndAnalyzeData(sampleCsv, outputJson, config);
    console.log("Processing results:", JSON.stringify(results, null, 2));
    
    // Clean up test files
    fs.unlinkSync(sampleCsv);
    if (fs.existsSync(outputJson)) {
      fs.unlinkSync(outputJson);
    }
  } catch (error) {
    console.error("Test error:", error);
  }
} 

File: python/large_function.py
"""
Python test file with a large function.

This file contains a large function that will later be broken down
into smaller functions to test DiffScope's ability to detect
function extraction and composition.
"""

import os
import json
import csv
import math
import statistics
from typing import List, Dict, Any, Optional, Union, Tuple


def process_and_analyze_data(
    input_path: str,
    output_path: str,
    config: Dict[str, Any]
) -> Dict[str, Any]:
    """
    Process data from input file, analyze it, and write results to output file.
    
    This function handles the entire data processing pipeline in one large function.
    It reads data from various file formats, performs analysis, generates statistics,
    handles errors, and writes the results to a specified output file.
    
    Args:
        input_path: Path to input file (CSV, JSON, or TXT)
        output_path: Path to output file (JSON)
        config: Configuration dictionary with processing options
        
    Returns:
        Dictionary with processing results and metadata
    """
    results = {
        "status": "success",
        "input_file": input_path,
        "output_file": output_path,
        "records_processed": 0,
        "errors": [],
        "warnings": [],
        "statistics": {}
    }
    
    # Validate inputs
    if not os.path.exists(input_path):
        results["status"] = "error"
        results["errors"].append(f"Input file does not exist: {input_path}")
        return results
    
    if not isinstance(config, dict):
        results["status"] = "error"
        results["errors"].append("Configuration must be a dictionary")
        return results
    
    # Extract configuration options
    column_mapping = config.get("column_mapping", {})
    filter_criteria = config.get("filter_criteria", {})
    aggregation_fields = config.get("aggregation_fields", [])
    include_stats = config.get("include_statistics", True)
    output_format = config.get("output_format", "json")
    
    # Determine file type from extension
    _, extension = os.path.splitext(input_path)
    extension = extension.lower()
    
    # Read and parse the input file
    data = []
    try:
        if extension == '.csv':
            with open(input_path, 'r', newline='') as csvfile:
                reader = csv.DictReader(csvfile)
                for row in reader:
                    data.append(row)
            results["file_type"] = "csv"
        elif extension == '.json':
            with open(input_path, 'r') as jsonfile:
                data = json.load(jsonfile)
            results["file_type"] = "json"
        elif extension == '.txt':
            with open(input_path, 'r') as txtfile:
                lines = txtfile.readlines()
                header = lines[0].strip().split(',')
                for line in lines[1:]:
                    values = line.strip().split(',')
                    if len(values) == len(header):
                        row = {header[i]: values[i] for i in range(len(header))}
                        data.append(row)
                    else:
                        results["warnings"].append(f"Skipped malformed line: {line.strip()}")
            results["file_type"] = "txt"
        else:
            results["status"] = "error"
            results["errors"].append(f"Unsupported file extension: {extension}")
            return results
    except Exception as e:
        results["status"] = "error"
        results["errors"].append(f"Error reading input file: {str(e)}")
        return results
    
    # Apply column mapping if provided
    if column_mapping:
        mapped_data = []
        for item in data:
            mapped_item = {}
            for old_key, new_key in column_mapping.items():
                if old_key in item:
                    mapped_item[new_key] = item[old_key]
                else:
                    # Copy unmapped fields
                    for k, v in item.items():
                        if k not in column_mapping:
                            mapped_item[k] = v
            mapped_data.append(mapped_item)
        data = mapped_data
    
    # Apply filtering if criteria provided
    if filter_criteria:
        filtered_data = []
        for item in data:
            include = True
            for field, criteria in filter_criteria.items():
                if field in item:
                    value = item[field]
                    
                    # Try to convert numeric strings to numbers for comparison
                    try:
                        if isinstance(value, str) and (value.isdigit() or 
                                                      (value.replace('.', '', 1).isdigit() and value.count('.') <= 1)):
                            value = float(value) if '.' in value else int(value)
                    except ValueError:
                        pass
                    
                    # Apply different types of criteria
                    if isinstance(criteria, dict):
                        if "eq" in criteria and value != criteria["eq"]:
                            include = False
                            break
                        if "ne" in criteria and value == criteria["ne"]:
                            include = False
                            break
                        if "gt" in criteria and not (isinstance(value, (int, float)) and value > criteria["gt"]):
                            include = False
                            break
                        if "lt" in criteria and not (isinstance(value, (int, float)) and value < criteria["lt"]):
                            include = False
                            break
                        if "in" in criteria and value not in criteria["in"]:
                            include = False
                            break
                        if "contains" in criteria and not (isinstance(value, str) and criteria["contains"] in value):
                            include = False
                            break
                    else:
                        # Simple equality check
                        if value != criteria:
                            include = False
                            break
            
            if include:
                filtered_data.append(item)
        
        results["filtered_count"] = len(filtered_data)
        results["original_count"] = len(data)
        data = filtered_data
    
    # Process numeric fields and calculate aggregations
    if data and aggregation_fields:
        aggregated_data = {}
        
        for field in aggregation_fields:
            numeric_values = []
            
            for item in data:
                if field in item:
                    value = item[field]
                    
                    # Try to convert string to number
                    if isinstance(value, str):
                        try:
                            value = float(value) if '.' in value else int(value)
                        except ValueError:
                            continue
                    
                    if isinstance(value, (int, float)):
                        numeric_values.append(value)
            
            if numeric_values:
                aggregated_data[field] = {
                    "count": len(numeric_values),
                    "sum": sum(numeric_values),
                    "average": sum(numeric_values) / len(numeric_values),
                    "min": min(numeric_values),
                    "max": max(numeric_values),
                }
                
                # Add more statistics if more than one value
                if len(numeric_values) > 1:
                    aggregated_data[field].update({
                        "median": statistics.median(numeric_values),
                        "stddev": statistics.stdev(numeric_values) if len(numeric_values) > 1 else 0,
                        "variance": statistics.variance(numeric_values) if len(numeric_values) > 1 else 0
                    })
        
        results["aggregations"] = aggregated_data
    
    # Compute overall statistics
    if include_stats and data:
        field_types = {}
        for item in data:
            for key, value in item.items():
                if key not in field_types:
                    field_types[key] = {"type": type(value).__name__, "count": 1}
                else:
                    field_types[key]["count"] += 1
        
        results["statistics"]["field_types"] = field_types
        results["statistics"]["record_count"] = len(data)
        results["statistics"]["field_count"] = len(field_types)
    
    # Prepare output
    output_data = {
        "metadata": results,
        "data": data
    }
    
    # Write to output file
    try:
        if output_format.lower() == "json":
            with open(output_path, 'w') as out_file:
                json.dump(output_data, out_file, indent=2)
        else:
            results["status"] = "error"
            results["errors"].append(f"Unsupported output format: {output_format}")
            return results
    except Exception as e:
        results["status"] = "error"
        results["errors"].append(f"Error writing output file: {str(e)}")
        return results
    
    results["records_processed"] = len(data)
    return results


if __name__ == "__main__":
    # Test with sample data
    config = {
        "column_mapping": {"name": "full_name", "age": "years"},
        "filter_criteria": {"years": {"gt": 20}},
        "aggregation_fields": ["years", "score"],
        "include_statistics": True,
        "output_format": "json"
    }
    
    # Create a sample CSV file for testing
    sample_csv = "tests/sample_data.csv"
    with open(sample_csv, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile)
        writer.writerow(["name", "age", "score", "active"])
        writer.writerow(["Alice", "25", "95", "true"])
        writer.writerow(["Bob", "30", "85", "true"])
        writer.writerow(["Charlie", "20", "75", "false"])
    
    output_json = "tests/output_data.json"
    results = process_and_analyze_data(sample_csv, output_json, config)
    print("Processing results:", json.dumps(results, indent=2))
    
    # Clean up test files
    os.remove(sample_csv)
    if os.path.exists(output_json):
        os.remove(output_json) 